okta/resource_okta_app_signon_policy_assignment.go

```diff
@@ -0,0 +1,326 @@
+	_, err := r.findAppSDKInnerResponse(ctx, state.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("update failed to find app %q for policy assignment", state.AppID.ValueString()),
+			err.Error(),
+		)
+		return
+	}
+
+	if !plan.PolicyID.Equal(state.PolicyID) {
+		// policy id has changed in the config, update
+		appID := plan.AppID.ValueString()
+		_, err = r.oktaSDKClientV3.ApplicationPoliciesApi.AssignApplicationPolicy(ctx, appID, plan.PolicyID.ValueString()).Execute()
+		if err != nil {
+			resp.Diagnostics.AddError(
+				fmt.Sprintf("couldn't re-assign policy %q to app %q", plan.PolicyID.ValueString(), appID),
+				err.Error(),
+			)
+			return
+		}
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (r *appSignonPolicyAssignmentResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	tflog.Warn(ctx, "True delete for an okta_app_signon_policy_assignment is a no-op as this resource will not delete an app or a policy. Additionally there is not an API endpoint to remove a sign on policy from an app, only update.")
+}
+
+func (r *appSignonPolicyAssignmentResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func (r *appSignonPolicyAssignmentResource) findAppSDKInnerResponse(ctx context.Context, appID string) (*okta.ListApplications200ResponseInner, error) {
+	appInnerResp, _, err := r.oktaSDKClientV3.ApplicationApi.GetApplication(ctx, appID).Execute()
+	return appInnerResp, err
+}
+
+func concreteAppID(src *okta.ListApplications200ResponseInner) (string, error) {
+	if src == nil {
+		return "", errors.New("list application inner response is nil")
+	}
+
+	if src.AutoLoginApplication != nil {
+		return *src.AutoLoginApplication.Id, nil
+	}
+
+	if src.BasicAuthApplication != nil {
+		return *src.BasicAuthApplication.Id, nil
+	}
+
+	if src.BookmarkApplication != nil {
+		return *src.BookmarkApplication.Id, nil
+	}
+
+	if src.BrowserPluginApplication != nil {
+		return *src.BrowserPluginApplication.Id, nil
+	}
+
+	if src.OpenIdConnectApplication != nil {
+		return *src.OpenIdConnectApplication.Id, nil
+	}
+
+	if src.SamlApplication != nil {
+		return *src.SamlApplication.Id, nil
+	}
+
+	if src.SecurePasswordStoreApplication != nil {
+		return *src.SecurePasswordStoreApplication.Id, nil
+	}
+
+	if src.WsFederationApplication != nil {
+		return *src.WsFederationApplication.Id, nil
+	}
+
+	return "", errors.New("list application inner response does not contain a concrete app")
+}
+
+func concreteAppLinks(src *okta.ListApplications200ResponseInner) (*okta.ApplicationLinks, error) {
+	if src == nil {
+		return nil, errors.New("list application inner response is nil")
+	}
+
+	if src.AutoLoginApplication != nil {
```

Comment:
fyi you can use scr.AutoLoginApplication.GetLinks() or GetLinksOK() to avoid manually check src.AutoLoginApplication != nil.

```// GetLinks returns the Links field value if set, zero value otherwise.
func (o *Application) GetLinks() ApplicationLinks {
	if o == nil || o.Links == nil {
		var ret ApplicationLinks
		return ret
	}
	return *o.Links
}

// GetLinksOk returns a tuple with the Links field value if set, nil otherwise
// and a boolean to check if the value has been set.
func (o *Application) GetLinksOk() (*ApplicationLinks, bool) {
	if o == nil || o.Links == nil {
		return nil, false
	}
	return o.Links, true
}
```

Reply:
Sadly, you still have to check if the application type is nil, or not. I like that get ok convenience pattern though.

I see a get actual instance, I'll have to see if I can make use of that.

`if id, ok = src.AutoLoginApplication.GetIdOk(); ok {`

```
TF_ACC=1 go test -tags unit -mod=readonly -test.v -run ^TestAccResourceOktaAppSignonPolicyAssignment$ ./okta 2>&1
=== RUN   TestAccResourceOktaAppSignonPolicyAssignment
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x2113cde]

goroutine 870 [running]:
github.com/okta/terraform-provider-okta/okta.concreteAppID(0xc000453758?)
        /Users/mikemondragon/projects/src/github.com/okta/terraform-provider-okta/okta/resource_okta_app_signon_policy_assignment.go:255 +0x1e
github.com/okta/terraform-provider-okta/okta.(*appSignonPolicyAssignmentResource).Create(0xc000472a28, {0x2ada6a8, 0xc00088aed0}, {{{{0x2adfeb0, 0xc00088b560}, {0x25d0740, 0xc00088b4d0}}, {0x2ae1ca0, 0xc00063a870}}, {{{0x2adfeb0, ...}, ...}, ...}, ...}, ...)
        /Users/mikemondragon/projects/src/github.com/okta/terraform-provider-okta/okta/resource_okta_app_signon_policy_assignment.go:104 +0x19d
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0xc000196f20, {0x2ada6a8, 0xc00088aed0}, 0xc000a69280, 0xc000a69220)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.3.5/internal/fwserver/server_createresource.go:101 +0x578
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0xc00069d3d8?, {0x2ada6a8, 0xc00088aed0}, 0xc0004532c0, 0xc000a693d8)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.3.5/internal/fwserver/server_applyresourcechange.go:57 +0x4a5
github.com/hashicorp/terraform-plugin-framework/internal/proto5server.(*Server).ApplyResourceChange(0xc000196f20, {0x2ada6a8?, 0xc00088ae10?}, 0xc000452f50)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.3.5/internal/proto5server/server_applyresourcechange.go:55 +0x3e5
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ApplyResourceChange({0xc0003d69f0, 0xc000751020, 0xc0003d6a20, 0xc0003d6a50, {0xc000245080, 0x2, 0x2}}, {0x2ada6a8?, 0xc00088ab70?}, 0xc000452f50)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-mux@v0.11.2/tf5muxserver/mux_server_ApplyResourceChange.go:30 +0x127
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0006ffe00, {0x2ada6a8?, 0xc00088a2d0?}, 0xc00016ea80)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/tf5server/server.go:821 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x278dba0?, 0xc0006ffe00}, {0x2ada6a8, 0xc00088a2d0}, 0xc00016ea10, 0x0)
        /Users/mikemondragon/projects/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:422 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000288960, {0x2ae0760, 0xc0006036c0}, 0xc000638120, 0xc0008d6810, 0x37bf140, 0x0)
        /Users/mikemondragon/projects/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:1337 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc000288960, {0x2ae0760, 0xc0006036c0}, 0xc000638120, 0x0)
        /Users/mikemondragon/projects/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:1714 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /Users/mikemondragon/projects/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:959 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 615
        /Users/mikemondragon/projects/pkg/mod/google.golang.org/grpc@v1.56.1/server.go:957 +0x165
FAIL    github.com/okta/terraform-provider-okta/okta    6.074s
FAIL
```

Looking at the unmarshal json code in v3 okta-sdk-golang I can see that it leaves unused application types as nil.
```go
// Unmarshal JSON data into one of the pointers in the struct  CUSTOM
func (dst *ListApplications200ResponseInner) UnmarshalJSON(data []byte) error {
	var err error
	// use discriminator value to speed up the lookup
	var jsonDict map[string]interface{}
	err = newStrictDecoder(data).Decode(&jsonDict)
	if err != nil {
		return fmt.Errorf("Failed to unmarshal JSON into map for the discriminator lookup.")
	}

	// check if the discriminator value is 'AUTO_LOGIN'
	if jsonDict["signOnMode"] == "AUTO_LOGIN" {
		// try to unmarshal JSON data into AutoLoginApplication
		err = json.Unmarshal(data, &dst.AutoLoginApplication)
		if err == nil {
			return nil // data stored in dst.AutoLoginApplication, return on the first match
		} else {
			dst.AutoLoginApplication = nil
			return fmt.Errorf("Failed to unmarshal ListApplications200ResponseInner as AutoLoginApplication: %s", err.Error())
		}
	}
```